### DISCUSSION TITLE
Search, select, download dependencies

### DISCUSSION THREAD

**Discussion by nipafx** [jbangdev/jbang#1813] [Ideas]
Before I get into it: I haven't really used JBang and what I am asking for may very well already exist (although a cursory search through the docs didn't unearth anything) or make no sense at all as a feature. Bear with me. 😃 

I often write small things in Java and doing that in a simple editor with just multi source-file execution is really nice. I don't really need tools for that (hence my lack of experience with JBang) and enjoy the simplicity of not using any. Until I need dependencies, of course. 🙈  I know JBang can download dependencies that are declared with `//DEPS`, but then my project is bound to JBang. Wouldn't it be nice if I could use just download a few JARs and then keep walking my tool-less path? 😁

Unrelated, I've recently written a small terminal app that executes a search on a specific website, parses the HTML, and displays the results as a series of lines. I can select the line I care about with arrow keys and start the download with _Enter_.

You can probably see where this is going: Wouldn't it be cool if there were a tool that I start in my project's `libs` folder (or wherever), use to search Maven, select one of the search results, and that then downloads the dependency tree and dumps it all into the folder? I was about to get started on this when I realized that there may be a project (this one! 😃) that could either already have the feature or include it in the future.

What do you think?

**Comment by maxandersen** [comment]
Not sure I fully get what you are after - are you asking for getting all the deps into a dir you can run? Then look at `jbang export`. 

If you are looking for usecase where you already have a set of jars and want to reverse lookup then that's probably more something like https://github.com/maveniverse/mima that could do that ? 

JBoss tools had the functionality so definitely possible.

**Reply by quintesse** [reply]
Like @maxandersen already mentioned (and if I understand your question correctly) the `jbang export` should indeed be able to do what you want, eg `jbang export portable G:A:V` for example would copy that artifact to the local directory while copying all its dependencies to a `lib` folder.

Btw, one of the advantages of the `//DEPS` lines is that they are just comments so you won't be "bound to JBang", it would just be an easy way to run the code, if you don't want to use JBang you can still copy the artifacts yourself :-)

**Reply by maxandersen** [reply]
Yeah good point @quintesse - wether you add deps and let jbang one time download or let jbang or some other tool download based on existing jars in a folder - are equal "dependent on".

**Reply by nipafx** [reply]
> are you asking for getting all the deps into a dir you can run

Basically, but also "yes and no". :D Yes, I'm looking to download dependencies of a script I'm developing. But no, those are not dependencies of "a jbang script" (i.e. not of a script that uses jbang's special comments to document/configure dependencies).

>  eg `jbang export portable G:A:V` for example would copy that artifact to the local directory while copying all its dependencies to a lib folder.

I just tried that and it mostly does what I want, although it leads to unexpected duplication:

```shell
$ mkdir test
$ cd test
$ jbang export portable org.jsoup:jsoup:1.17.2
[jbang] Updating jar...
[jbang] Exported to /home/nipa/code/test/jsoup.jar
$ tree
.
├── jsoup.jar
└── lib
    └── jsoup-1.17.2.jar
```

> so you won't be "bound to JBang"

"bound to JBang" as in "I can't run it without JBang". 😉 I'm looking for a tool that I only need to use once per dependency (update) not once per project execution. But I understood Max' follow-up comment to propose that I could use `//DEPS` comments to declare dependencies and then immediately download them with JBang into a local folder, from which point on I don't need JBang until the next dependency.

**Reply by maxandersen** [reply]
ok so coming back to this - there is something that is 90% what you are asking for...but it does result in a .jar file you can just ignore.

Lets say your project has a `main.java` file then just do this:

`jbang export portable --deps=org.jsoup:jsoup:1.17.2 main.java`

You can also just have that main.java list out the deps using jbang syntax:

<h5>main.java</h5>
```java 
//DEPS org.jsoup:jsoup:1.17.2 stub.build
```

then its just:

`jbang export portable main.java` 

But then of course the .java file with have traces of jbang.

Anyway jbang is flexible - it can be visible or you can use it fully undercover :)

In either case you get:

```
├── lib
│   └── jsoup-1.17.2.jar
├── main.java
└── main.jar
```

you can delete the main.jar file as it has no use to you.

For now jbang export assumes you are trying to make "something" runnable where something is a jar, gav or source files.
I had a chat with @quintesse and jbang has all the mechanics to NOT require specifying something on export - allowing to export basically just an "environment". Its similar to what interactive mode does; it just setups a classpath that you as a user starts jshell with.

We can probably make that work if you find the deletion of the main.jar too much hassle.

wdyt? useful to you?

p.s. I found a bug in `jbang export` that makes it error if there are files in the lib folder already  - I'll fix that ASAP. workaround is to rm the lib folder first.

**Reply by maxandersen** [reply]
PR for the bug: https://github.com/jbangdev/jbang/pull/1821/files

feature request for the export portable without a script/target: https://github.com/jbangdev/jbang/issues/1822

**Reply by maxandersen** [reply]
1.117.1 is available with the bug fixed.

**Reply by nipafx** [reply]
Hi Max, thank you very much for looking into this. I'll experiment with these options next time I need them. :)

**Comment by nipafx** [comment]
A reader of this discussion pointed me towards `jbang gavsearch@jbangdev $name`:

```
$ jbang gavsearch@jbangdev jsoup
> Searching for `jsoup` on search.maven.org...
> [maven, gradle, jbang]
> ? coordinates org.jsoup:jsoup:1.18.1
> maven:
> <dependency>
>   <groupId>org.jsoup</groupId>
>   <artifactId>jsoup</artifactId>
>   <version>1.18.1</version>
> </dependency>
> gradle:
> org.jsoup:jsoup:1.18.1
> jbang:
> //DEPS org.jsoup:jsoup:1.18.1
> ? Action
>   Copy maven to Clipboard
>   Copy gradle to Clipboard
> ❯ Copy jbang to Clipboard
>   Search older versions
>   Quit
```

Why didn't you tell me, @maxandersen?! 😂 

So it's not exactly what I was thinking, but it's (a) very close and (b) the last remaining piece of the puzzle. If we slightly reorganize existing JBang features, we get exactly what I was looking for:

* move the "input box" for the search term into the gavsearch
* add "Download JAR" and "Download JAR with dependencies" to the menu
* connect those to JBang's already existing dependency download capabilities

What do you think, could that be an interesting feature?

**Reply by maxandersen** [reply]
sorry - missed this thread ;/ 

moving input box for the search item into gavsearch would make sense (just have it popup if you dont specify a search.

download jar/download jar with dependencies - sure.

..maybe something for maveniverse/toolbox to do ? 

not sure when I get to try but gavsearch code is at https://github.com/jbangdev/jbang-catalog/blob/main/gavsearch.java and quite simple

**Reply by quintesse** [reply]
In the meantime I created this project that focuses on npm-like behaviour: https://github.com/codejive/java-jpm

Easiest way to use it is still using JBang :-)

I've also added some fixes to the upstream JLines3 library that gavsearch and java-jpm use and I'm working on some more fixes/features that will make the experience even better. (The JLine3 library has some unfortunate limitations that I'm trying to fix)
